txpool/txpool_test.go

```diff
@@ -1,1125 +1,1701 @@
+				assert.NoError(t, err)
+			}()
+			pool.handleEnqueueRequest(<-pool.enqueueReqCh)
 		}
 
-		for _, testCase := range testTable {
-			transaction, _ := pool.Pop()
+		// verify all 10 are enqueued
+		assert.Equal(t, uint64(10), pool.accounts.get(addr1).enqueued.length())
+		assert.Equal(t, uint64(0), pool.accounts.get(addr1).promoted.length())
 
-			assert.NotNil(t, transaction)
+		// execute the promotion handler
+		pool.handlePromoteRequest(req)
 
-			actual := TestCase{
-				From:     transaction.From,
-				Gas:      transaction.Gas,
-				GasPrice: transaction.GasPrice,
-				Nonce:    transaction.Nonce,
-				Value:    transaction.Value,
-			}
+		assert.Equal(t, uint64(10), pool.gauge.read())
+		assert.Equal(t, uint64(10), pool.accounts.get(addr1).getNonce())
 
-			assert.EqualValues(t, testCase, actual)
+		assert.Equal(t, uint64(0), pool.accounts.get(addr1).enqueued.length())
+		assert.Equal(t, uint64(10), pool.accounts.get(addr1).promoted.length())
+	})
+
+	t.Run("one tx -> one promotion", func(t *testing.T) {
+		/* In this scenario, each received tx will be instantly promoted.
+		All txs are sent in the order of expected nonce. */
+		pool, err := newTestPool()
+		assert.NoError(t, err)
+		pool.SetSigner(&mockSigner{})
+		pool.EnableDev()
+
+		for nonce := uint64(0); nonce < 20; nonce++ {
+			go func(nonce uint64) {
+				err := pool.addTx(local, newDummyTx(addr1, nonce, 1))
+				assert.NoError(t, err)
+			}(nonce)
+			go pool.handleEnqueueRequest(<-pool.enqueueReqCh)
+			pool.handlePromoteRequest(<-pool.promoteReqCh)
 		}
 
-		empty, _ := pool.Pop()
-		assert.Nil(t, empty)
-	}
+		assert.Equal(t, uint64(20), pool.gauge.read())
+		assert.Equal(t, uint64(20), pool.accounts.get(addr1).getNonce())
+
+		assert.Equal(t, uint64(0), pool.accounts.get(addr1).enqueued.length())
+		assert.Equal(t, uint64(20), pool.accounts.get(addr1).promoted.length())
+	})
+
+	t.Run("promote demoted tx with low nonce", func(t *testing.T) {
+		pool, err := newTestPool()
+		assert.NoError(t, err)
+		pool.SetSigner(&mockSigner{})
+		pool.EnableDev()
+
+		{ // setup prestate
+			acc := pool.createAccountOnce(addr1)
+			acc.setNonce(7)
+		}
+
+		// send recovered tx
+		go pool.handleEnqueueRequest(enqueueRequest{
+			tx:      newDummyTx(addr1, 4, 1),
+			demoted: true,
+		})
+
+		// promote
+		pool.handlePromoteRequest(<-pool.promoteReqCh)
+
+		assert.Equal(t, uint64(0), pool.accounts.get(addr1).enqueued.length())
+		assert.Equal(t, uint64(1), pool.accounts.get(addr1).promoted.length())
+	})
+}
 
-	t.Run("the higher priced transaction should be popped first", func(t *testing.T) {
-		test(t, []TestCase{
+func TestResetAccount(t *testing.T) {
+	t.Run("reset promoted", func(t *testing.T) {
+		testCases := []struct {
+			name     string
+			txs      transactions
+			newNonce uint64
+			expected result
+		}{
 			{
-				From:     addr1,
-				Gas:      validGasLimit,
-				GasPrice: big.NewInt(2),
-				Value:    big.NewInt(0),
+				name: "prune all txs",
```

Comment:
Please consider improving these test names. For example, "prune all txs from the promoted queue with nonce less than new nonce" or something similar which is more descriptive and provides more information in case when we see this test failing. This could be improved in other places but I will leave only this comment here.

Reply:
[Renamed](*https://github.com/0xPolygon/polygon-sdk/pull/287/commits/bae5c576e85356ce5aa9d9963a52214d7cfbe8e7)